### PR TITLE
[SPARK-23957][SQL] Sorts in subqueries are redundant and can be removed

### PR DESCRIPTION
**Pull request by dilipbiswal**
## What changes were proposed in this pull request?
Thanks to @henryr for the original idea at https://github.com/apache/spark/pull/21049

Description from the original PR :
Subqueries (at least in SQL) have 'bag of tuples' semantics. Ordering
them is therefore redundant (unless combined with a limit). 

This patch removes the top sort operators from the subquery plans.

This closes https://github.com/apache/spark/pull/21049.

## How was this patch tested?
Added test cases in SubquerySuite to cover in, exists and scalar subqueries.

Please review http://spark.apache.org/contributing.html before opening a pull request.

### CODE REVIEW

**Comment by gatorsmile** [comment]
cc @maryannxue

File: sql/core/src/test/scala/org/apache/spark/sql/SubquerySuite.scala [hunk 1]
```diff
@@ -970,4 +973,300 @@ class SubquerySuite extends QueryTest with SharedSQLContext {
...
+          |              GROUP BY t2.c1
+          |              HAVING   count(*) >= 1
+          |              ORDER BY max(t2.c1)
+          |              LIMIT 1)
+        """.stripMargin
+      assert(getNumSortsInQuery(query5) == 1)
+    }
+  }
 }
+
```
Comment (maropu) [inline sql/core/src/test/scala/org/apache/spark/sql/SubquerySuite.scala]: super nit: remove this blank line

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala [hunk 2]
```diff
@@ -164,10 +164,20 @@ abstract class Optimizer(sessionCatalog: SessionCatalog)
    * Optimize all the subqueries inside expression.
    */
   object OptimizeSubqueries extends Rule[LogicalPlan] {
+    private def removeTopLevelSorts(plan: LogicalPlan): LogicalPlan = {
```
Comment (maropu) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala]: nit: `removeTopLevelSort`? (I think this func removes a single sort on the top?)

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala [hunk 3]
```diff
@@ -164,10 +164,20 @@ abstract class Optimizer(sessionCatalog: SessionCatalog)
...
+        case Project(fields, child) => Project(fields, removeTopLevelSorts(child))
+        case other => other
+      }
+    }
     def apply(plan: LogicalPlan): LogicalPlan = plan transformAllExpressions {
       case s: SubqueryExpression =>
         val Subquery(newPlan) = Optimizer.this.execute(Subquery(s.plan))
-        s.withNewPlan(newPlan)
+        // At this point we have an optimized subquery plan that we are going to attach
+        // to this subquery expression. Here we can safely remove any top level sorts
```
Comment (maropu) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala]: super nit: `any top level sort`?

File: sql/core/src/test/scala/org/apache/spark/sql/SubquerySuite.scala [hunk 4]
```diff
@@ -970,4 +973,300 @@ class SubquerySuite extends QueryTest with SharedSQLContext {
...
+      val query2 =
+        """
+           |SELECT c1
+           |FROM   t1
+           |WHERE  EXISTS (SELECT c1
+           |               FROM   (SELECT *
+           |                      FROM   t2
+           |                      WHERE t2.c1 = t1.c1
+           |                      ORDER  BY t2.c2) t2
+           |              ORDER  BY t2.c1)
```
Comment (maropu) [inline sql/core/src/test/scala/org/apache/spark/sql/SubquerySuite.scala]: super nit: add one space before `ORDER`

**Comment by maropu** [comment]
LGTM except for minor comments

**Comment by maropu** [comment]
Also, could you add `Closes #21049` in the description?

**Comment by maropu** [comment]
retest this please

**Comment by gatorsmile** [comment]
Ideally, this is a perfect fix. We can make it more general to remove all the unnecessary sorts during the query planning. However, this optimization is still nice to have since the sorts removed by this PR are not rare.

**Comment by gatorsmile** [comment]
LGTM

Thanks! Merged to master.

**Comment by dilipbiswal** [comment]
Thank you very much @gatorsmile and @maropu

**Pull request closed without merging**